### PR TITLE
[FLINK-2124] [streaming] Fix behavior of FromElementsFunction when T …

### PR DESCRIPTION
**Pull request by jreiffers**
…isn't Serializable

rewrite FromElementsFunction to work with arbitrary types. The elements are serialized to a byte array (using a TypeSerializer) when the FromElementsFunction is constructed and deserialized when run is called.

I also removed the (Collection<T>) constructor, because all Collections are Iterable.

### CODE REVIEW

File: flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/source/FromElementsFunction.java [hunk 1]
```diff
@@ -17,37 +17,71 @@
...
+		ByteArrayInputStream bais = new ByteArrayInputStream(elements);
+		DataInputView input = new InputViewDataInputStreamWrapper(new DataInputStream(bais));
 
-		while (isRunning && it.hasNext()) {
-			ctx.collect(it.next());
+		while (isRunning && bais.available() > 0) {
+			value = serializer.deserialize(value, input);
+			ctx.collect(value);
 		}
+		// closing the DataOutputStream would just close the ByteArrayInputStream, which doesn't do anything
```
Comment (tillrohrmann) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/source/FromElementsFunction.java]: I guess you meant the `DataInputStream`.

File: flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/source/FromElementsFunction.java [hunk 2]
```diff
@@ -17,37 +17,71 @@
...
 
-	public FromElementsFunction(Iterable<T> elements) {
-		this.iterable = elements;
+	public FromElementsFunction(TypeSerializer<T> serializer, Iterable<T> elements) {
+		ByteArrayOutputStream baos = new ByteArrayOutputStream();
+		OutputViewDataOutputStreamWrapper wrapper = new OutputViewDataOutputStreamWrapper(new DataOutputStream(baos));
+
+		try {
+			for (T element : elements)
+				serializer.serialize(element, wrapper);
```
Comment (tillrohrmann) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/source/FromElementsFunction.java]: Implicit coding conventions are to put curly brackets also around one line `for` statements.

File: flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/source/FromElementsFunction.java [hunk 3]
```diff
@@ -17,37 +17,71 @@
...
 package org.apache.flink.streaming.api.functions.source;
 
-import java.util.Arrays;
-import java.util.Collection;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.core.memory.DataInputView;
+import org.apache.flink.core.memory.InputViewDataInputStreamWrapper;
+import org.apache.flink.core.memory.OutputViewDataOutputStreamWrapper;
+
+import java.io.*;
```
Comment (tillrohrmann) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/source/FromElementsFunction.java]: The checkstyle plugin complains about the `.*` import here. You have to import the different streams explicitly.

**Comment by tillrohrmann** [comment]
Thanks for your contribution @jreiffers. The PR looks really good :-) There is only a problem with the `.*` import.

I was thinking about adding a test case where we create a `FromElementsFunction` with non serializable data elements. What do you think?

**Comment by tillrohrmann** [comment]
LGTM. +1 for merging once Travis has given green light.

File: flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/source/FromElementsFunction.java [hunk 4]
```diff
@@ -17,37 +17,81 @@
...
+	public FromElementsFunction(TypeSerializer<T> serializer, Iterable<T> elements) {
+		ByteArrayOutputStream baos = new ByteArrayOutputStream();
+		OutputViewDataOutputStreamWrapper wrapper = new OutputViewDataOutputStreamWrapper(new DataOutputStream(baos));
+
+		try {
+			for (T element : elements) {
+				serializer.serialize(element, wrapper);
+			}
+		} catch (IOException e) {
+			// ByteArrayOutputStream doesn't throw IOExceptions when written to
```
Comment (rmetzger) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/source/FromElementsFunction.java]: I think not forwarding the exception here is a bad idea.

File: flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/source/FromElementsFunction.java [hunk 5]
```diff
@@ -17,37 +17,81 @@
...
+					@Override
+					public T next() {
+						return elements[index++];
+					}
 
-	public FromElementsFunction(Collection<T> elements) {
-		this.iterable = elements;
+					@Override
+					public void remove() {
+						throw new UnsupportedOperationException();
```
Comment (mbalassi) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/source/FromElementsFunction.java]: Please add some explanation message to the exception. On paper it is for intrnal use, but still. :)

File: flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/source/FromElementsFunction.java [hunk 6]
```diff
@@ -17,37 +17,81 @@
...
+	private final TypeSerializer<T> serializer;
+	private final byte[] elements;
 
 	private volatile boolean isRunning = true;
 
-	public FromElementsFunction(T... elements) {
-		this.iterable = Arrays.asList(elements);
-	}
+	public FromElementsFunction(TypeSerializer<T> serializer, final T... elements) {
+		this(serializer, new Iterable<T>() {
```
Comment (StephanEwen) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/source/FromElementsFunction.java]: I think you can make this much simpler by using `Arrays.asList(elements)`

File: flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/source/FromElementsFunction.java [hunk 7]
```diff
@@ -17,37 +17,81 @@
...
+		// closing the DataOutputStream would just flush the ByteArrayOutputStream, which in turn doesn't do anything.
+
+		this.serializer = serializer;
+		this.elements = baos.toByteArray();
 	}
 
 	@Override
 	public void run(SourceContext<T> ctx) throws Exception {
-		Iterator<T> it = iterable.iterator();
+		T value = serializer.createInstance();
```
Comment (StephanEwen) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/source/FromElementsFunction.java]: Let's not use object reusing mode by default. It is not guaranteed to always work (some abstract types are not instantiable) and should only work on user-requested activation. Use `value = serializer.deserialize(input);` instead.

**Comment by StephanEwen** [comment]
I think a good addition would be to remember the number of elements in the set, and use that to validate that the deserialization worked properly. Serializers may be user-defined (when they want) and the user-defined serializers may be buggy. Counting the number of elements and verifying that everything is consumed after deserialization is a good way of figuring out whether things are sane, and giving an error if not.

File: flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/source/FromElementsFunction.java (hunk 6 above)
Comment (jreiffers) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/source/FromElementsFunction.java]: Maybe. Arrays.asList also unnecessarily copies the array.

File: flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/source/FromElementsFunction.java (hunk 4 above)
Comment (jreiffers) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/source/FromElementsFunction.java]: Right, I missed that.

File: flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/source/FromElementsFunction.java (hunk 6 above)
Comment (jreiffers) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/source/FromElementsFunction.java]: As it turns out, it actually doesn't... the more you know!

**Pull request closed without merging**